x/gamm/keeper/pool.go

```diff
@@ -105,97 +105,110 @@ func (k Keeper) DeletePool(ctx sdk.Context, poolId uint64) error {
 //
-// All locks on this pool share must be unlocked prior to execution. Use LockupKeeper.ForceUnlock
+// CONTRACT: All locks on this pool share must be unlocked prior to execution. Use LockupKeeper.ForceUnlock
 // on remaining locks before calling this function.
-// func (k Keeper) CleanupBalancerPool(ctx sdk.Context, poolIds []uint64, excludedModules []string) (err error) {
-// 	pools := make(map[string]types.PoolI)
-// 	totalShares := make(map[string]sdk.Int)
-// 	for _, poolId := range poolIds {
-// 		pool, err := k.GetPool(ctx, poolId)
-// 		if err != nil {
-// 			return err
-// 		}
-// 		shareDenom := pool.GetTotalShares().Denom
-// 		pools[shareDenom] = pool
-// 		totalShares[shareDenom] = pool.GetTotalShares().Amount
-// 	}
-
-// 	moduleAccounts := make(map[string]string)
-// 	for _, module := range excludedModules {
-// 		moduleAccounts[string(authtypes.NewModuleAddress(module))] = module
-// 	}
-
-// 	// first iterate through the share holders and burn them
-// 	k.bankKeeper.IterateAllBalances(ctx, func(addr sdk.AccAddress, coin sdk.Coin) (stop bool) {
-// 		if coin.Amount.IsZero() {
-// 			return
-// 		}
-
-// 		pool, ok := pools[coin.Denom]
-// 		if !ok {
-// 			return
-// 		}
-
-// 		// track the iterated shares
-// 		pool.SubTotalShares(coin.Amount)
-// 		pools[coin.Denom] = pool
-
-// 		// check if the shareholder is a module
-// 		if _, ok = moduleAccounts[coin.Denom]; ok {
-// 			return
-// 		}
-
-// 		// Burn the share tokens
-// 		err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, addr, types.ModuleName, sdk.Coins{coin})
-// 		if err != nil {
-// 			return true
-// 		}
-
-// 		err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, sdk.Coins{coin})
-// 		if err != nil {
-// 			return true
-// 		}
-
-// 		// Refund assets
-// 		for _, asset := range pool.GetAllPoolAssets() {
-// 			// lpShareEquivalentTokens = (amount in pool) * (your shares) / (total shares)
-// 			lpShareEquivalentTokens := asset.Token.Amount.Mul(coin.Amount).Quo(totalShares[coin.Denom])
-// 			if lpShareEquivalentTokens.IsZero() {
-// 				continue
-// 			}
-// 			err = k.bankKeeper.SendCoins(
-// 				ctx, pool.GetAddress(), addr, sdk.Coins{{asset.Token.Denom, lpShareEquivalentTokens}})
-// 			if err != nil {
-// 				return true
-// 			}
-// 		}
-
-// 		return false
-// 	})
-
-// 	if err != nil {
-// 		return err
-// 	}
-
-// 	for _, pool := range pools {
-// 		// sanity check
-// 		if !pool.GetTotalShares().IsZero() {
-// 			panic("pool total share should be zero after cleanup")
-// 		}
-
-// 		err = k.DeletePool(ctx, pool.GetId())
-// 		if err != nil {
-// 			return err
-// 		}
-// 	}
-
-// 	return nil
-// }
+func (k Keeper) CleanupBalancerPool(ctx sdk.Context, poolIds []uint64) (err error) {
+	// we use maps here because we can't alter the state of pool directly
+	poolAddresses := make(map[string]sdk.AccAddress)
+	totalShares := make(map[string]sdk.Int)
+	poolLiquidity := make(map[string]sdk.Coins)
```

Comment:
Wdyt about having:
```
type poolInfo struct {
    address sdk.AccAddress\
    totalShares sdk.Int
    liquidity sdk.Coins
}

poolInfo := map[string]poolInfo{}
```

It seems to be a common pattern to need more than one of this field on each iteration so by consolidating this in a struct, we would do one map lookup instead of 3

Reply:
We could get all 3 at the beginning of iteration and update the map with the new amounts at the end to minimize the number of map accesses.

Reply:
makes sense to me! fixed!